File: articles/waves-oscillations-and-resonance/python/resonance_parameter_sweep.py
"""
Resonance Parameter Sweep

This workflow computes the steady-state amplitude of a driven damped oscillator:

    A(omega) = (F0/m) / sqrt((omega0^2 - omega^2)^2 + (2 gamma omega)^2)
"""

from pathlib import Path

import numpy as np
import pandas as pd


def response_amplitude(
    mass_kg: float,
    spring_constant_n_per_m: float,
    damping_coefficient_kg_per_s: float,
    driving_force_n: float,
    driving_angular_frequency_rad_per_s: np.ndarray,
) -> np.ndarray:
    """
    Compute steady-state driven oscillator amplitude.
    """
    omega0 = np.sqrt(spring_constant_n_per_m / mass_kg)
    gamma = damping_coefficient_kg_per_s / (2.0 * mass_kg)

    denominator = np.sqrt(
        (omega0**2 - driving_angular_frequency_rad_per_s**2) ** 2
        + (2.0 * gamma * driving_angular_frequency_rad_per_s) ** 2
    )

    return (driving_force_n / mass_kg) / denominator


def main() -> None:
    """
    Generate resonance curves for several damping values.
    """
    article_dir = Path(__file__).resolve().parents[1]
    output_path = article_dir / "data" / "computed_resonance_parameter_sweep.csv"
    summary_path = article_dir / "data" / "computed_resonance_peak_summary.csv"

    mass_kg = 1.0
    spring_constant_n_per_m = 25.0
    driving_force_n = 1.0
    omega0 = np.sqrt(spring_constant_n_per_m / mass_kg)

    damping_values = [0.2, 0.6, 1.2]
    omega_values = np.linspace(0.1, 12.0, 400)

    rows = []

    for damping in damping_values:
        amplitudes = response_amplitude(
            mass_kg,
            spring_constant_n_per_m,
            damping,
            driving_force_n,
            omega_values,
        )

        for omega, amplitude in zip(omega_values, amplitudes):
            rows.append(
                {
                    "damping_coefficient_kg_per_s": damping,
                    "driving_angular_frequency_rad_per_s": omega,
                    "frequency_ratio": omega / omega0,
                    "response_amplitude_m": amplitude,
                }
            )

    table = pd.DataFrame(rows)

    summary = (
        table.sort_values("response_amplitude_m", ascending=False)
        .groupby("damping_coefficient_kg_per_s")
        .head(1)
        .sort_values("damping_coefficient_kg_per_s")
        .reset_index(drop=True)
    )

    table.to_csv(output_path, index=False)
    summary.to_csv(summary_path, index=False)

    print("Resonance peak summary:")
    print(summary.round(8).to_string(index=False))
    print(f"\nSaved sweep to: {output_path}")
    print(f"Saved summary to: {summary_path}")


if __name__ == "__main__":
    main()

File: articles/waves-oscillations-and-resonance/cpp/resonance_parameter_sweep.cpp
/*
Resonance Parameter Sweep

This C++ workflow computes driven oscillator response amplitude:

    A(omega) = (F0/m) / sqrt((omega0^2 - omega^2)^2 + (2 gamma omega)^2)
*/

#include <cmath>
#include <iomanip>
#include <iostream>
#include <vector>

double response_amplitude(
    double mass_kg,
    double spring_constant,
    double damping,
    double driving_force,
    double omega
) {
    double omega0 = std::sqrt(spring_constant / mass_kg);
    double gamma = damping / (2.0 * mass_kg);

    double denominator = std::sqrt(
        std::pow(omega0 * omega0 - omega * omega, 2.0)
        + std::pow(2.0 * gamma * omega, 2.0)
    );

    return (driving_force / mass_kg) / denominator;
}

int main() {
    const double mass_kg = 1.0;
    const double spring_constant = 25.0;
    const double driving_force = 1.0;

    std::vector<double> damping_values = {0.2, 0.6, 1.2};

    std::cout << "damping_kg_per_s,omega_rad_per_s,response_amplitude_m\n";

    for (double damping : damping_values) {
        for (double omega = 0.1; omega <= 12.0001; omega += 0.1) {
            double amplitude = response_amplitude(
                mass_kg,
                spring_constant,
                damping,
                driving_force,
                omega
            );

            std::cout << std::setprecision(12)
                      << damping << ","
                      << omega << ","
                      << amplitude << "\n";
        }
    }

    return 0;
}
